File: typescript/lld/booking/library/FineStrategy.ts
import { Loan } from './Loan';

// Strategy: overdue fine rule is pluggable.
export interface FineStrategy {
    compute(loan: Loan, returnedAt: number): number;
}

export class PerDayFine implements FineStrategy {
    constructor(private readonly perDay: number) {}
    compute(loan: Loan, returnedAt: number): number {
        const overdueMs = returnedAt - loan.dueAt;
        if (overdueMs <= 0) return 0;
        return Math.ceil(overdueMs / 86_400_000) * this.perDay;   // ms -> days
    }
}

File: java/lld/booking/library/FineStrategy.java
// Strategy: overdue fine rule is pluggable.
public interface FineStrategy {
    double compute(Loan loan, long returnedAt);
}

class PerDayFine implements FineStrategy {
    private final double perDay;
    public PerDayFine(double perDay) { this.perDay = perDay; }
    public double compute(Loan loan, long returnedAt) {
        long overdue = returnedAt - loan.dueAt;
        if (overdue <= 0) return 0;
        return (long) Math.ceil(overdue / 86_400_000.0) * perDay;
    }
}